server/src/main/java/org/elasticsearch/cluster/routing/ExpectedShardSizeEstimator.java

```diff
@@ -51,7 +51,12 @@ public static long getExpectedShardSize(
         } else if (shard.unassigned() && shard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
             return snapshotShardSizeInfo.getShardSize(shard, defaultValue);
         } else {
-            return clusterInfo.getShardSize(shard, defaultValue);
+            var shardSize = clusterInfo.getShardSize(shard.shardId(), shard.primary());
+            if (shardSize == null && shard.primary() == false) {
+                // derive replica size from corresponding primary
+                shardSize = clusterInfo.getShardSize(shard.shardId(), true);
```

Comment:
According to a failing test in CI, it seems that sometimes the shard size of the primary may be -1 (e.g., when INITIALIZING it seems), which may not be appropriate. So you may think about assigning the value only if it's >0, else it stays null.